pkg/epp/datalayer/factory.go

```diff
@@ -44,3 +48,79 @@ type EndpointFactory interface {
 	NewEndpoint(parent context.Context, inpod *corev1.Pod, poolinfo PoolInfo) Endpoint
 	ReleaseEndpoint(ep Endpoint)
 }
+
+// EndpointLifecycle manages the life cycle (creation and termination) of
+// endpoints.
+type EndpointLifecycle struct {
+	sources         []DataSource  // data sources for collectors
+	collectors      sync.Map      // collectors map. key: Pod namespaced name, value: *Collector
+	refreshInterval time.Duration // metrics refresh interval
+}
+
+// NewEndpointFactory returns a new endpoint for factory, managing collectors for
+// its endpoints.
+// Sources might not be known at creation time (e.g., loaded from configuration
+// file), so they're not passed at this point.
+func NewEndpointFactory(sources []DataSource, refreshMetricsInterval time.Duration) *EndpointLifecycle {
+	lc := &EndpointLifecycle{
+		sources:         make([]DataSource, len(sources)),
+		collectors:      sync.Map{},
+		refreshInterval: refreshMetricsInterval,
+	}
+	_ = copy(lc.sources, sources)
+
+	return lc
```

Comment:
we're under the assumption that data sources are initialized only at bootstrap and that this function is called after all data sources are already known.. 
can we remove the `copy` call and just init the data sources from the arg?
e.g.,:
```suggestion
	return &EndpointLifecycle{
		sources:         sources,
		collectors:      sync.Map{},
		refreshInterval: refreshMetricsInterval,
	}
```

Reply:
we can remove the copy under the assumption that there shall be no changes to the original slice.